tests/nested/classic/hotplug/task.yaml

```diff
@@ -0,0 +1,259 @@
+    . "$TESTSLIB/nested.sh"
+
+    if execute_remote "udevadm info -e" | MATCH "ID_MODEL.QEMU_USB_SERIAL"; then
+        echo "USB serial already registered, exiting..."
+        exit 1
+    fi
+
+    execute_remote "sudo snap install hello-world"
+
+    echo "Enabling hotplug"
+    execute_remote "sudo snap set core experimental.hotplug=true"
+
+    echo "Plugging the device"
+    hotplug_add_dev1
+
+    # sanity checks to make sure qemu setup is correct
+    for _ in $(seq 5); do
+        if execute_remote "udevadm info -e" | MATCH "ID_MODEL.QEMU_USB_SERIAL"; then
+            break
+        fi
+        sleep 1
+    done
+
+    execute_remote "ls /dev/tty*" | MATCH "ttyUSB0"
+
+    echo "Checking that qemuusbserial hotplug slot is present"
+    check_slot_present qemuusbserial
+    check_slot_not_gone qemuusbserial
+    check_slot_device_path qemuusbserial "/dev/ttyUSB0"
+
+    echo "Unplugging the device with no connections on the slot"
+    hotplug_del_dev1
+
+    # sanity check to make sure qemu event was triggered correctly
+    for _ in $(seq 5); do
+        if execute_remote "udevadm info -e" | MATCH "ID_MODEL.QEMU_USB_SERIAL"; then
+            sleep 1
+        else
+            break
+        fi
+    done
+    if execute_remote "udevadm info -e" | MATCH "ID_MODEL.QEMU_USB_SERIAL"; then
+        echo "USB serial should not be registered anymore, exiting..."
+        exit 1
+    fi
+
+    echo "Checking that hotplug slot is no longer present"
+    check_slot_not_present qemuusbserial
+    check_slot_not_present_in_state qemuusbserial
+
+    echo "Plugging the device back"
+    hotplug_add_dev1
+
+    echo "Checking that qemuusbserial hotplug slot is present again"
+    check_slot_present qemuusbserial
+
+    echo "Installing test snap with serial port plug"
+    execute_remote "sudo snap install --dangerous serial-port-hotplug_1.0_all.snap"
+
+    echo "Connecting hotplug slot of the first device"
+    execute_remote "sudo snap connect serial-port-hotplug:serial-port :qemuusbserial"
+    check_slot_connected qemuusbserial
+
+    echo "Veryfing serial-port permissions of the snap"
+    verify_apparmor_profile "/dev/ttyUSB0"
+    execute_remote "/snap/bin/serial-port-hotplug.consumer" | MATCH "Access to /dev/ttyUSB0 ok"
+    execute_remote "/snap/bin/serial-port-hotplug.consumer" | MATCH "Access to /dev/ttyUSB1 failed"
+
+    echo "Unplugging the device"
+    hotplug_del_dev1
+
+    echo "Checking that qemuusbserial hotplug slot is no longer present, but is remembered"
+    check_slot_not_present qemuusbserial
+    check_slot_gone qemuusbserial
+
+    echo "Plugging device back, checking that the slot and connection are restored"
+    hotplug_add_dev1
+    check_slot_connected qemuusbserial
+    check_slot_not_gone qemuusbserial
+
+    echo "Unplugging the device again"
+    hotplug_del_dev1
+
+    echo "Checking that qemuusbserial hotplug slot is not present, but is still remembered"
+    check_slot_not_present qemuusbserial
```

Comment:
The test is failing on this check, this is some debug info from the vm that could help to understand.
https://paste.ubuntu.com/p/zxSQzKYypB/
I just added some debug info to the test to understand why the test could fail https://paste.ubuntu.com/p/dssWvyH5Xy/

Reply:
Thanks, I added your debug snippet to this test, with one small addition.
As discussed earlier, I pushed some fixes and the test should be passing now, at least on 16.04 and 18.04,